File: java/src/main/java/org/craftedsw/tripservicekata/trip/TripService.java
package org.craftedsw.tripservicekata.trip;

import java.util.ArrayList;
import java.util.List;

import org.craftedsw.tripservicekata.exception.UserNotLoggedInException;
import org.craftedsw.tripservicekata.user.User;
import org.craftedsw.tripservicekata.user.UserSession;

public class TripService {

    private SecurityService securityService;

    public TripService() {
        this(new SecurityService());
    }

    public TripService(SecurityService securityService) {
        this.securityService = securityService;
    }

    public List<Trip> getTripsByUser(User user) throws UserNotLoggedInException {
        User loggedUser = securityService.getLoggedInUser();
        if (loggedUser == null) {
            throw new UserNotLoggedInException();
        }

        boolean isFriend = user.isFriend(loggedUser);

        List<Trip> tripList = new ArrayList<>();
        if (isFriend) {
            tripList = findTripsByUser(user);
        }
        return tripList;
    }

    protected List<Trip> findTripsByUser(User user) {
        return TripDAO.findTripsByUser(user);
    }

}

class SecurityService {

    protected User getLoggedInUser() {
        return UserSession.getInstance().getLoggedUser();
    }

}

File: python-objects/TripService.py
from TripDAO import TripDAO
from UserNotLoggedInException import UserNotLoggedInException
from UserSession import UserSession


class TripService(object):
    def get_trips_by_user(self, user):

        def user_session_get_instance():
            return UserSession.get_instance()

        def trip_dao_find_trips(user):
            TripDAO.find_trips_by_user(user)

        self._get_trips_pure(user, user_session_get_instance, trip_dao_find_trips)

    def _get_trips_pure(self, user, user_session_get_instance, trip_dao_find_trips):
        logged_user = user_session_get_instance().get_logged_user()
        if not logged_user:
            raise UserNotLoggedInException()

        is_friend = self.is_friend_with(logged_user, user)

        if is_friend:
            return trip_dao_find_trips(user)
        else:
            trip_list = []
            return trip_list

    def is_friend_with(self, logged_user, user):
        for friend in user.get_friends():
            if friend is logged_user:
                return True
        return False
